### DISCUSSION TITLE
Can't use async in redirect: from GoRouter

### DISCUSSION THREAD

**Discussion by MaximeGer** [csells/go_router#299] [Q&A] [answered]
I use the package flutter_secure_storage to store a jwt. And to handle if the jwt is expired I have to run asynchronous code. To redirect to the login page I need to use the redirect: of the GoRouter so it can checks everytime a new screen is displayed. 

The problem is that redirect: from GoRouter wants a String? and not a Future<String?> so I can't use async and await keywords otherwise I'll get my screen full of red underlined text. I tried using a ".then" callback but without success for the moment cause it doesn't await the end of the callback.

**Comment by csells** [comment]
The `redirect` function itself can't return an async value but you can deal with async values in a standard way in Flutter. Check out this sample: https://gorouter.dev/async-data

**Reply by MaximeGer** [reply]
I see. A possible solution would be to use another route that will check the jwt and according to parameters will redirect to the asked screen.
So instead of using the redirect: option, I could use a custom JwtCheckScreen and make my rules of redirection in it. 
Like for example : (pseudo-code) if jwt expires then context.go('/login') else context.go(params.["page-url"]).
With params.["page-url"] being a string sent in the parameters of the route.

I would lose the ability to use the combo redirect + refreshListenable but it might work. If you have a better solution in mind, I'm listening.  

Edit : Now that I think about it, it would be visually really bad to use an intermediary page each time you change a Screen. There would be 2 animations for each change of Screen. 
Really looking forward for a better solution

**Comment by csells** [comment]
You can use the combo of redirect + refreshListener by updating the object that you pass to refreshListener be updated with that data. I do a similar thing in [the loading page example](https://github.com/csells/go_router/blob/main/go_router/example/lib/loading_page.dart).

**Reply by MaximeGer** [reply]
But is there a way to avoid using an intermediary loading screen ? Or at least make it invisible for the user ? 
Like I said in my previous reply, I would prefer not having 2 new page animations each time I access one new page

**Comment by csells** [comment]
Like any async data in Flutter, you can load it and show it however you like. Here're some docs you might find useful: https://gorouter.dev/async-data

**Comment by MaximeGer** [comment, marked as answer]
I found a solution and made a  [repo](https://github.com/Zealyy/flutter_nodejs_postgresql_jwt_authentication) with all my code so it can help others. 
The solution I found is to forget redirect / refreshListenable and to use a widget that I wrap on every routes : [the widget](https://github.com/Zealyy/flutter_nodejs_postgresql_jwt_authentication/blob/main/mobile/lib/widgets/JwtChecker.dart) and [the main.dart with the router](https://github.com/Zealyy/flutter_nodejs_postgresql_jwt_authentication/blob/main/mobile/lib/main.dart)

Edit: the widget is a FutureBuilder
